### PR TITLE
Fix false statement about 'let' hoisting

### PR DESCRIPTION
**Pull request by feimosi**
See the link I've provided in the patch for more info.

### CODE REVIEW

**Comment by GoToLoop** [comment]
If both `let` & `const` declared variables can't in no way be accessed before the statement they're declared at, for a programmer's point-of-view they're effectively never hoisted, regardless their actual internal implementation! :shipit:

**Comment by feimosi** [comment]
There's a significant difference in whether something is hoisted or not. Currently if you try to access or override that variable you're gonna get a `ReferenceError`, which some developers might not expect. I recommend reading the link provided :)

**Comment by metagrover** [comment]
yes, the article is certainly insightful. thanks @feimosi 👍

**Comment by GoToLoop** [comment]
Let me rephrase it: As human programmers, whether or not `let` & `const` actually hoist like `var`, that is whether the variables created by them happen at the very beginning within a function, is completely irrelevant.

The perceived effect for us is that they're created exactly where they're declared at; and cease to exist once their scope finishes.

Which by the way is very diff. compared to `var` (and even inner `function` declarations), which can be accessed from the very beginning of its function.

Who knows other languages like Java and even C the same thing happens? That all variables are actually pre-created & hoisted at the top of their respective functions, before their statement lines?

However, no 1 learning Java or C has ever needed to trouble themselves to know such internal implementations in order to program in them!
Therefore why JS community is complicating things by mentioning esoteric terms such as "Temporal Dead Zone"?

Wouldn't be much clearer simply stating that `let`, `const` & `class` have the same expected behavior as any other language?
And that `var` & `function` declarations are the only 1s which have the "deviate" hoisting behavior, which is apparently exclusive to JS language?

**Comment by feimosi** [comment]
It's not that simple as you may think. 

First of all, I've corrected the false phrase that `let` & `const` are hoisted, which is simply not true. It's against the spec, you could as well say that functions in JS are not objects or `class` keyword implements real object oriented paradigm - just because in your opinion, to XX% of people that goes unnoticed.

How can we expect to have high qualified engineers in the community, when we teach them from unreliable resources? I think at some point of everyone's career it's necessary to dive into 'implementation' details of the language (which is IMHO not TDZ), that's how you become a senior developer.

Secondly, it's not `the same expected behavior as any other language`, although it tries to emulate it, there are slight differences. Let me use a classic example here:

``` js
if (true) {
  let a = 1;
  if (true) {
    // Can't access a from the upper scope - ReferenceError
    let a = 2;
  }
}
```

This is a completely valid snippet, while in Java it's not possible to override variables that way. I'm not a Java expert, but similar code yields:
`Compilation error: variable a is already defined in method main(String[])`

Of course it's not that important in this tutorial to write whole paragraphs about TDZ - that's why I've included a link for those interested.

**Comment by GoToLoop** [comment]
> First of all, I've corrected the false phrase that let & const are hoisted, ...

That's the phrase you requested and was successfully pulled in this repository some hours ago.

Previous: No Hoisting takes place when we use `let` keyword.
Yours: Both `let` and `const` are hoisted, ...

It shoulda been: newest `let`, `const` & `class` aren't hoisted like the old `var` & `function` declarations.

For the 3rd time: Whether all variables are actually created at the beginning of a function call or not is implementation's detail, which can vary among the many JS engines btW, and not whatsoever the ECMA's definition for the JS language.

For all JS code, the behavior for the newest keywords `let`, `const` & `class` is not hoisted + block scoped.

**Comment by feimosi** [comment]
I'm not here to argue. I've expressed my opinion with the reasoning behind it and your arguments don't convince me in any way. I feel like you're missing the point. Anyway the author has decided.

Have a nice day, cheers :)

**Comment by metagrover** [comment]
Hey guys, I have updated the statement about 'let' hoisting. I feel it's much better now.

**Comment by GoToLoop** [comment]
It's got better now. Although I still think it's much simpler stating they don't hoist at all, like all normal languages. BtW, I spot you mention `function` there too. You should mention `class` as well.

In short, there are 5 variable declaration keywords in JS: `var`, `function`, `let`, `const` & `class`.
The 1st 2 keywords hoist, while the last ES6 3 functionally don't.

**Comment by metagrover** [comment]
There is a difference, class declarations are not hoisted. Read here - https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes

**Comment by GoToLoop** [comment]
> ... `class` declarations are not hoisted.

I thought I've already stated that 3 times already!  :expressionless:

> Wouldn't be much clearer simply stating that let, const & class have the same expected behavior...
> 
> ... the behavior for the newest keywords let, const & class aren't hoisted + block scoped.
> 
> ... while the last ES6 3 functionally don't.

Besides I'm the only 1 who remembered `class` keyword is also a variable declaration here.  :bowtie:

**Comment by metagrover** [comment]
Ohh my bad. Poor choice of words. I meant to say that `let` and `const` hoist but `class` declarations don't. Also, I have mentioned:

> Class declarations are not hoisted. You first need to declare your class and then access it, otherwise ReferenceError will be thrown.

under 13. Classes in ES6 section

Anyway, the current statement clarifies things much better. Let's put this to rest. ✌️ 😄

**Comment by GoToLoop** [comment]
OK, this is a very long discussion now. But just 1 little extra observation.
That explanation about keyword `class` can be applied equally for `let` & `const`:  :wink:  

> Let/Const declarations are not hoisted. You first need to declare your variable and then access it, otherwise ReferenceError will be thrown.

A very interested link about at http://StackOverflow.com/questions/31219420/are-variables-declared-with-let-or-const-not-hoisted-in-es6
